File: creme/stats/ptp.py
from . import base
from . import maximum
from . import minimum


class PeakToPeak(base.Univariate):
    """Running peak to peak (max - min).

    Attributes:
        max (stats.Max): The running max.
        min (stats.Min): The running min.
        p2p (float): The running peak to peak.

    Example:

        ::

            >>> from creme import stats

            >>> X = [1, -4, 3, -2, 2, 4]
            >>> ptp = stats.PeakToPeak()
            >>> for x in X:
            ...     print(ptp.update(x).get())
            0
            5
            7
            7
            7
            8

    """

    def __init__(self):
        self.max = maximum.Max()
        self.min = minimum.Min()

    @property
    def name(self):
        return 'ptp'

    def update(self, x):
        self.max.update(x)
        self.min.update(x)
        return self

    def get(self):
        return self.max.get() - self.min.get()


class RollingPeakToPeak(base.RollingUnivariate):
    """Running peak to peak (max - min) over a window.

    Parameters:
        window_size (int): Size of the rolling window.

    Attributes:
        max (stars.RollingMax): The running rolling max.
        min (stars.RollingMin): The running rolling min.

    Example:

        ::

            >>> from creme import stats

            >>> X = [1, -4, 3, -2, 2, 1]
            >>> ptp = stats.RollingPeakToPeak(window_size=2)
            >>> for x in X:
            ...     print(ptp.update(x).get())
            0
            5
            7
            5
            4
            1

    """

    def __init__(self, window_size):
        self.max = maximum.RollingMax(window_size)
        self.min = minimum.RollingMin(window_size)

    @property
    def window_size(self):
        return self.max.window_size

    def update(self, x):
        self.max.update(x)
        self.min.update(x)
        return self

    def get(self):
        return self.max.get() - self.min.get()

File: _watermill/src/ptp.rs
use crate::maximum::{Max, RollingMax};
use crate::minimum::{Min, RollingMin};
use crate::stats::Univariate;
use num::{Float, FromPrimitive};
use serde::{Deserialize, Serialize};
use std::ops::{AddAssign, SubAssign};
/// Running peak to peak (max - min).
/// # Examples
/// ```
/// use watermill::ptp::PeakToPeak;
/// use watermill::stats::Univariate;
/// let mut running_peak_to_peak: PeakToPeak<f64> = PeakToPeak::new();
/// for i in 1..10{
///     running_peak_to_peak.update(i as f64);
/// }
/// assert_eq!(running_peak_to_peak.get(), 8.0);
/// ```
///
#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub struct PeakToPeak<F: Float + FromPrimitive + AddAssign + SubAssign> {
    pub min: Min<F>,
    pub max: Max<F>,
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Default for PeakToPeak<F> {
    fn default() -> Self {
        Self {
            min: Min::new(),
            max: Max::new(),
        }
    }
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> PeakToPeak<F> {
    pub fn new() -> Self {
        Self::default()
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for PeakToPeak<F> {
    fn update(&mut self, x: F) {
        self.min.update(x);
        self.max.update(x);
    }
    fn get(&self) -> F {
        self.max.get() - self.min.get()
    }
}

/// Rolling peak to peak (max - min).
/// # Arguments
/// * `window_size` - Size of the rolling window.
/// # Examples
/// ```
/// use watermill::ptp::RollingPeakToPeak;
/// use watermill::stats::Univariate;
/// let mut rolling_peak_to_peak: RollingPeakToPeak<f64> = RollingPeakToPeak::new(3);
/// for i in 1..10{
///     rolling_peak_to_peak.update(i as f64);
/// }
/// assert_eq!(rolling_peak_to_peak.get(), 2.0);
/// ```
///
#[derive(Serialize, Deserialize)]
pub struct RollingPeakToPeak<F: Float + FromPrimitive + AddAssign + SubAssign> {
    pub min: RollingMin<F>,
    pub max: RollingMax<F>,
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> RollingPeakToPeak<F> {
    pub fn new(window_size: usize) -> Self {
        Self {
            min: RollingMin::new(window_size),
            max: RollingMax::new(window_size),
        }
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for RollingPeakToPeak<F> {
    fn update(&mut self, x: F) {
        self.min.update(x);
        self.max.update(x);
    }
    fn get(&self) -> F {
        self.max.get() - self.min.get()
    }
}
